distributed/scheduler.py

```diff
@@ -4609,27 +4614,38 @@ async def add_nanny(self, comm: Comm, address: str) -> None:
                 self._starting_nannies.discard(address)
                 self._starting_nannies_cond.notify_all()
 
-    def _match_graph_with_tasks(
+    def _find_lost_dependencies(
         self,
         dsk: dict[Key, T_runspec],
         dependencies: dict[Key, set[Key]],
         keys: set[Key],
     ) -> set[Key]:
-        n = -1
         lost_keys = set()
-        while len(dsk) != n:  # walk through new tasks, cancel any bad deps
-            n = len(dsk)
-            for k, deps in list(dependencies.items()):
-                if (k not in self.tasks and k not in dsk) or any(
-                    dep not in self.tasks and dep not in dsk for dep in deps
-                ):  # bad key
-                    lost_keys.add(k)
-                    logger.info("User asked for computation on lost data, %s", k)
-                    dsk.pop(k, None)
-                    del dependencies[k]
-                    if k in keys:
-                        keys.remove(k)
-                del deps
+        seen: set[Key] = set()
+        sadd = seen.add
+        for k in list(keys):
+            work = {k}
+            wpop = work.pop
+            wupdate = work.update
+            while work:
+                d = wpop()
+                if d in seen:
+                    continue
+                sadd(d)
+                if d not in dsk:
+                    if d not in self.tasks:
+                        lost_keys.add(d)
+                        lost_keys.add(k)
+                        logger.info("User asked for computation on lost data, %s", k)
+                        dependencies.pop(d, None)
+                        keys.discard(k)
+                    continue
+                wupdate(dsk[d].dependencies)
+        return lost_keys
```

Comment:
I rewrote this section. I had issues with it and just barely understand the old code (and ran into multiple bugs in the recent past)